File: console/src/quickstarts/regional_cache/quickstart.py
from __future__ import annotations

import json
import os
import time
from typing import Any

from epoch_sdk import (
    RegionalCacheClient,
    RegionalCacheExpectation,
    RegionalCacheLockGuard,
    RegionalCacheMultiplexMutation,
    RegionalCacheMutation,
    RegionalCacheTransform,
    RegionalCacheValue,
    RegionalScope,
)


def result(document: dict[str, Any]) -> dict[str, Any]:
    operation = document["receipt"]["outcome"]["result"]
    assert isinstance(operation, dict)
    return operation


client = RegionalCacheClient(
    os.getenv(
        "EPOCH_REGIONAL_ENDPOINTS",
        "http://127.0.0.1:18661,http://127.0.0.1:18662,http://127.0.0.1:18663",
    ).split(","),
    token=os.getenv("EPOCH_TOKEN", "epoch-dev-admin-v1"),
    scope=RegionalScope("acme", "shop", "dev", "core"),
    timeout=3.0,
)

written = client.set(
    "sessions",
    0,
    "docs-python-cache-set-v1",
    "profile",
    RegionalCacheValue.string("alice"),
)
replayed = client.set(
    "sessions",
    0,
    "docs-python-cache-set-v1",
    "profile",
    RegionalCacheValue.string("alice"),
)
version = int(result(written)["item"]["version"])
compared = client.compare_and_set(
    "sessions",
    0,
    "docs-python-cache-cas-v1",
    "profile",
    RegionalCacheExpectation.version(version),
    RegionalCacheValue.hash({"name": "alice", "role": "admin"}),
)
committed_get = client.get("sessions", 0, "docs-python-cache-get-v1", "profile")
observation = client.observe("sessions", 0, "profile")
revision = int(observation["observation"]["shard_revision"])
batch = client.atomic_batch(
    "sessions",
    0,
    "docs-python-cache-batch-v1",
    revision,
    [
        RegionalCacheMutation.set("visits", RegionalCacheValue.counter(1)),
        RegionalCacheMutation.set(
            "recent", RegionalCacheValue.list(["home", "checkout"])
        ),
        RegionalCacheMutation.set("roles", RegionalCacheValue.set(["admin", "buyer"])),
        RegionalCacheMutation.set(
            "rank", RegionalCacheValue.sorted_set({"alice": 9.5})
        ),
        RegionalCacheMutation.set("avatar", RegionalCacheValue.blob(b"epoch")),
    ],
)
acquired = client.acquire_lock(
    "sessions", 0, "docs-python-cache-lock-v1", "profile-lock", "docs-python", 1, 60_000
)
lease_token = result(acquired)["lease_token"]
guard = RegionalCacheLockGuard("profile-lock", "docs-python", 1, lease_token)
guarded = client.increment(
    "sessions",
    0,
    "docs-python-cache-guarded-increment-v1",
    "visits",
    1,
    lock_guard=guard,
)
released = client.release_lock(
    "sessions",
    0,
    "docs-python-cache-release-v1",
    "profile-lock",
    "docs-python",
    1,
    lease_token,
)
ephemeral = client.set(
    "sessions",
    0,
    "docs-python-cache-ttl-v1",
    "flash",
    RegionalCacheValue.string("short"),
    ttl_ms=1,
)
time.sleep(0.01)
maintained = client.maintain(
    "sessions", 0, "docs-python-cache-maintain-v1", max_expirations=100
)
cold = client.set(
    "sessions",
    0,
    "docs-python-cache-cold-v1",
    "profile-archive",
    RegionalCacheValue.string("alice-archive"),
    storage_class="cold",
)
backup = client.backup("sessions", 0)
captured_revision = int(backup["captured_revision"])
client.set(
    "sessions",
    0,
    f"docs-python-cache-restore-scratch-{captured_revision}",
    "restore-scratch",
    RegionalCacheValue.string("remove-me"),
)
restored = client.restore(
    "sessions",
    0,
    f"docs-python-cache-restore-{captured_revision}",
    backup["artifact_base64"],
    captured_revision,
)
bitmap = client.transform(
    "sessions",
    0,
    "docs-python-cache-bitmap-v1",
    "feature-flags",
    RegionalCacheTransform("bitmap_set", {"bit": 7, "value": True}),
)
bitmap_query = client.query(
    "sessions", 0, "bitmap_get", {"key": "feature-flags", "bit": 7}
)
multiplexed = client.multiplex(
    "sessions",
    0,
    [
        RegionalCacheMultiplexMutation(
            "presence",
            "docs-python-cache-multiplex-presence-v1",
            RegionalCacheMutation.set("presence", RegionalCacheValue.string("online")),
        ),
        RegionalCacheMultiplexMutation(
            "notifications",
            "docs-python-cache-multiplex-notifications-v1",
            RegionalCacheMutation.increment("notifications", 1),
        ),
    ],
)
subscription = client.create_subscription(
    "sessions", 0, channels=["session.audit"], patterns=["session.*"]
)
subscription_id = subscription["subscription_id"]
published = client.publish("sessions", 0, "session.audit", {"profile": "alice"})
polled = client.poll_subscription("sessions", 0, subscription_id, limit=10)
deleted_subscription = client.delete_subscription("sessions", 0, subscription_id)
changes = client.changes("sessions", 0, 1, limit=100)

print(
    json.dumps(
        {
            "set": written,
            "exact_retry": replayed,
            "cas": compared,
            "committed_get": committed_get,
            "atomic_batch": batch,
            "guarded_increment": guarded,
            "release": released,
            "ttl": ephemeral,
            "maintain": maintained,
            "cold": cold,
            "backup": backup,
            "restore": restored,
            "bitmap": bitmap,
            "bitmap_query": bitmap_query,
            "multiplex": multiplexed,
            "subscription": subscription,
            "publish": published,
            "poll": polled,
            "delete_subscription": deleted_subscription,
            "changes": changes,
            "profile": client.observe("sessions", 0, "profile"),
            "status": client.status("sessions", 0),
        },
        indent=2,
    )
)

File: console/src/quickstarts/regional_cache/quickstart.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"epoch.local/epoch/sdk/go/epoch"
)

func main() {
	client, err := epoch.NewRegionalCacheClient(
		strings.Split(environment("EPOCH_REGIONAL_ENDPOINTS", "http://127.0.0.1:18661,http://127.0.0.1:18662,http://127.0.0.1:18663"), ","),
		environment("EPOCH_TOKEN", "epoch-dev-admin-v1"),
		epoch.RegionalScope{Organization: "acme", Project: "shop", Environment: "dev", Namespace: "core"},
		3*time.Second,
	)
	must(err)
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
	defer cancel()

	written, err := client.Set(ctx, "sessions", 0, "docs-go-cache-set-v1", "profile", epoch.NewRegionalCacheString("alice"), epoch.RegionalCacheWriteOptions{})
	must(err)
	replayed, err := client.Set(ctx, "sessions", 0, "docs-go-cache-set-v1", "profile", epoch.NewRegionalCacheString("alice"), epoch.RegionalCacheWriteOptions{})
	must(err)
	version := decimal(result(written)["item"].(map[string]any)["version"])
	compared, err := client.CompareAndSet(ctx, "sessions", 0, "docs-go-cache-cas-v1", "profile", epoch.RegionalCacheVersion(version), epoch.NewRegionalCacheHash(map[string]string{"name": "alice", "role": "admin"}), epoch.RegionalCacheWriteOptions{})
	must(err)
	committedGet, err := client.Get(ctx, "sessions", 0, "docs-go-cache-get-v1", "profile")
	must(err)
	observation, err := client.Observe(ctx, "sessions", 0, "profile")
	must(err)
	revision := decimal(observation["observation"].(map[string]any)["shard_revision"])
	roles, err := epoch.NewRegionalCacheSet([]string{"admin", "buyer"})
	must(err)
	rank, err := epoch.NewRegionalCacheSortedSet(map[string]float64{"alice": 9.5})
	must(err)
	batch, err := client.AtomicBatch(ctx, "sessions", 0, "docs-go-cache-batch-v1", revision, []epoch.RegionalCacheMutation{
		epoch.NewRegionalCacheSetMutation("visits", epoch.NewRegionalCacheCounter(1), nil),
		epoch.NewRegionalCacheSetMutation("recent", epoch.NewRegionalCacheList([]string{"home", "checkout"}), nil),
		epoch.NewRegionalCacheSetMutation("roles", roles, nil),
		epoch.NewRegionalCacheSetMutation("rank", rank, nil),
		epoch.NewRegionalCacheSetMutation("avatar", epoch.NewRegionalCacheBlob([]byte("epoch")), nil),
	}, nil)
	must(err)
	acquired, err := client.AcquireLock(ctx, "sessions", 0, "docs-go-cache-lock-v1", "profile-lock", "docs-go", 1, 60_000)
	must(err)
	leaseToken := result(acquired)["lease_token"].(string)
	guard := &epoch.RegionalCacheLockGuard{LockKey: "profile-lock", Owner: "docs-go", OwnerEpoch: 1, LeaseToken: leaseToken}
	guarded, err := client.Increment(ctx, "sessions", 0, "docs-go-cache-guarded-increment-v1", "visits", 1, epoch.RegionalCacheIncrementOptions{LockGuard: guard})
	must(err)
	released, err := client.ReleaseLock(ctx, "sessions", 0, "docs-go-cache-release-v1", "profile-lock", "docs-go", 1, leaseToken)
	must(err)
	ttl := uint64(1)
	ephemeral, err := client.Set(ctx, "sessions", 0, "docs-go-cache-ttl-v1", "flash", epoch.NewRegionalCacheString("short"), epoch.RegionalCacheWriteOptions{TTLMS: &ttl})
	must(err)
	time.Sleep(10 * time.Millisecond)
	maintained, err := client.Maintain(ctx, "sessions", 0, "docs-go-cache-maintain-v1", 100)
	must(err)
	cold, err := client.Set(ctx, "sessions", 0, "docs-go-cache-cold-v1", "profile-archive", epoch.NewRegionalCacheString("alice-archive"), epoch.RegionalCacheWriteOptions{StorageClass: "cold"})
	must(err)
	backup, err := client.Backup(ctx, "sessions", 0)
	must(err)
	capturedRevision := decimal(backup["captured_revision"])
	_, err = client.Set(ctx, "sessions", 0, fmt.Sprintf("docs-go-cache-restore-scratch-%d", capturedRevision), "restore-scratch", epoch.NewRegionalCacheString("remove-me"), epoch.RegionalCacheWriteOptions{})
	must(err)
	restored, err := client.Restore(ctx, "sessions", 0, fmt.Sprintf("docs-go-cache-restore-%d", capturedRevision), backup["artifact_base64"].(string), capturedRevision)
	must(err)
	bitmapTransform, err := epoch.NewRegionalCacheTransform("bitmap_set", map[string]any{"bit": 7, "value": true})
	must(err)
	bitmap, err := client.Transform(ctx, "sessions", 0, "docs-go-cache-bitmap-v1", "feature-flags", bitmapTransform, nil, nil, nil)
	must(err)
	bitmapQuery, err := client.Query(ctx, "sessions", 0, map[string]any{"kind": "bitmap_get", "key": "feature-flags", "bit": 7})
	must(err)
	presence, err := epoch.NewRegionalCacheMultiplexMutation("presence", "docs-go-cache-multiplex-presence-v1", epoch.NewRegionalCacheSetMutation("presence", epoch.NewRegionalCacheString("online"), nil))
	must(err)
	notifications, err := epoch.NewRegionalCacheMultiplexMutation("notifications", "docs-go-cache-multiplex-notifications-v1", epoch.NewRegionalCacheIncrementMutation("notifications", 1, nil, nil))
	must(err)
	multiplexed, err := client.Multiplex(ctx, "sessions", 0, []epoch.RegionalCacheMultiplexMutation{presence, notifications})
	must(err)
	subscription, err := client.CreateSubscription(ctx, "sessions", 0, []string{"session.audit"}, []string{"session.*"})
	must(err)
	subscriptionID := subscription["subscription_id"].(string)
	published, err := client.Publish(ctx, "sessions", 0, "session.audit", map[string]any{"profile": "alice"})
	must(err)
	polled, err := client.PollSubscription(ctx, "sessions", 0, subscriptionID, 10)
	must(err)
	deletedSubscription, err := client.DeleteSubscription(ctx, "sessions", 0, subscriptionID)
	must(err)
	changes, err := client.Changes(ctx, "sessions", 0, 1, 100)
	must(err)
	status, err := client.Status(ctx, "sessions", 0)
	must(err)

	output, err := json.MarshalIndent(map[string]any{
		"set": written, "exact_retry": replayed, "cas": compared, "committed_get": committedGet, "atomic_batch": batch,
		"guarded_increment": guarded, "release": released, "ttl": ephemeral, "maintain": maintained,
		"cold": cold, "backup": backup, "restore": restored, "bitmap": bitmap, "bitmap_query": bitmapQuery,
		"multiplex": multiplexed, "subscription": subscription, "publish": published, "poll": polled,
		"delete_subscription": deletedSubscription, "changes": changes,
		"profile": observation, "status": status,
	}, "", "  ")
	must(err)
	fmt.Println(string(output))
}

func result(document epoch.Document) map[string]any {
	return document["receipt"].(map[string]any)["outcome"].(map[string]any)["result"].(map[string]any)
}

func decimal(value any) uint64 {
	parsed, err := strconv.ParseUint(fmt.Sprint(value), 10, 64)
	must(err)
	return parsed
}

func environment(name, fallback string) string {
	if value := strings.TrimSpace(os.Getenv(name)); value != "" {
		return value
	}
	return fallback
}

func must(err error) {
	if err != nil {
		panic(err)
	}
}
